src/ir/possible-contents.h

```diff
@@ -0,0 +1,513 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ir_possible_contents_h
+#define wasm_ir_possible_contents_h
+
+#include <variant>
+
+#include "ir/possible-constant.h"
+#include "ir/subtypes.h"
+#include "support/small_vector.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+//
+// PossibleContents represents the possible contents at a particular location
+// (such as in a local or in a function parameter). This is a little similar to
+// PossibleConstantValues, but considers more types of contents than constant
+// values - in particular, it can track types to some extent.
+//
+// The specific contents this can vary over are:
+//
+//  * None:            No possible value.
+//
+//  * Literal:         One possible constant value like an i32 of 42.
+//
+//  * Global:          The name of a global whose value is here. We do not know
+//                     the actual value at compile time, but we know it is equal
+//                     to that global. Typically we can only infer this for
+//                     immutable globals.
+//
+//  * ExactType:       Any possible value of a specific exact type - *not*
+//                     including subtypes. For example, (struct.new $Foo) has
+//                     ExactType contents of type $Foo.
+//                     If the type here is nullable then null is also allowed.
+//                     TODO: Add ConeType, which would include subtypes.
+//                     TODO: Add ExactTypePlusContents or such, which would be
+//                           used on e.g. a struct.new with an immutable field
+//                           to which we assign a constant: not only do we know
+//                           the exact type, but also certain field's values.
+//
+//  * Many:            Anything else. Many things are possible here, and we do
+//                     not track what they might be, so we must assume the worst
+//                     in the calling code.
+//
+// Note that we prefer not to emit ExactType{i32} etc., or any other type that
+// has no subtyping: instead, we emit Many for such things. While ExactType{i32}
+// would be correct, it adds no additional information: we know the type from
+// the wasm, and we know no subtype is possible, so we know it exactly.
+// Furthermore, using Many instead simplifies the logic in various places, in
+// particular, Many is a clear signal that we've hit the worst case and can't
+// do any better in that location, and can stop looking at it. If instead of
+// Many we emitted ExactType{i32} we'd need to also check for an exact type of
+// a type without subtypes, when trying to reason about hitting the worst case.
```

Comment:
Sorry I still don't really understand this paragraph.

> While ExactType{i32} would be correct, it adds no additional information: we know the type from the wasm, and we know no subtype is possible, so we know it exactly.

Yeah we know the type exactly, but I've never seen we query it in `Flower`. If a child gets `Many` type, it just stops flowing the value to its targets. Then isn't this loss of information?

> Furthermore, using Many instead simplifies the logic in various places, in particular, Many is a clear signal that we've hit the worst case and can't do any better in that location, and can stop looking at it. If instead of Many we emitted ExactType{i32} we'd need to also check for an exact type of a type without subtypes, when trying to reason about hitting the worst case.

Don't really get this part either. Why `ExactType{i32}` means we can't do any better? If we assign `ExactType` to a location, it will be propagated to its target locations, but we instead say it's `Many` and don't propagate anything anymore.

Maybe my misunderstanding stems from that I don't really know how this class is used for optimization. (I haven't looked at the GUFA PR yet) But I assumed the purpose of this analysis is to have more specific info on what can go to each location; correct me if I'm wrong.

Reply:
Is `ExactType` is supposed to only reference types? If so, would renaming this to `ExactRefType` be clearer?

Reply:
On the one hand, when we have a wasm location, like an `i32.add`, and we have `PossibleContents` for it, then `ExactType{i32}` adds nothing since the `i32.add` already has that type. That's why we don't use it in most places. However, on the other hand, there may be cases where we only have a `PossibleContents` but *not* have a wasm location for it. That's the case in our unit tests for `PossibleContents`, and it might make sense elsewhere. In such contexts I think `ExactType{i32}` is useful, as we do want a way to express that.

So overall, I think when optimizing we don't want `ExactType{i32}`, but in general we don't want to disallow it entirely.

Reply:
> Maybe my misunderstanding stems from that I don't really know how this class is used for optimization. (I haven't looked at the GUFA PR yet)

Yes, maybe it was not best to split up the PRs... this PR is much smaller but then the motivation is not present.

But hopefully my last comment in this thread is helpful? The key thing is that when we optimize we also have the context of the location that the content is for. And that location has a type. So if subtyping is not an issue, Many tells us no more than an exact type.

Reply:
> On the one hand, when we have a wasm location, like an `i32.add`, and we have `PossibleContents` for it, then `ExactType{i32}` adds nothing since the `i32.add` already has that type. That's why we don't use it in most places.

This looks the repetition of the comment above, so I have the same question from my previous comment:
Yeah we know the type exactly, but I've never seen we query it in `Flower`. If a child gets `Many` type, it just stops flowing the value to its targets. Then isn't this loss of information?

> However, on the other hand, there may be cases where we only have a `PossibleContents` but _not_ have a wasm location for it. That's the case in our unit tests for `PossibleContents`, and it might make sense elsewhere.

Can you point what test is this?

Reply:
> For example, see `declaredHeapType` and how we use it on cone reads: if we have `Many` there then we just fall back to the declared type in the wasm. (I think that answers your point `I've never seen we query it in Flower` - if not, perhaps I've misunderstood you?)

Ah I see, we do use it in `readFromData`. What I meant was I have not seen it being used in all the other places. For example, binary or unary operations. You mean the user of this `PossibleContents` class, which is presumably in the original GUFA PR uses it?

Reply:
Yes, fair point, it's just used there.

Ok, I did some experimentation and it seems like it's practical to do it as you suggest. I pushed that now. This is about the same amount of code as before, and I benchmarked it as being the same speed. And it is a simpler mental model, to not special-case exact types of non-references. So I agree your approach is best here.

This is done in https://github.com/WebAssembly/binaryen/pull/4685/commits/6c5affd482c96e7a1a076fb84fee86391c835b3f and then optimized in https://github.com/WebAssembly/binaryen/pull/4685/commits/ec3de0213b0bed1177add2a315c834476b7737cb - without that, it would be slightly slower, I measured - and comments are simplified in https://github.com/WebAssembly/binaryen/pull/4685/commits/2c3fb7827d1305022f25a2f2add7f048450e0b02

Reply:
It's not that I was insisting we do this; I was mostly asking why the current code is like this.. But anyway if this is clearer and doesn't have any performance hit that should be OK I guess.

But what this fixes is only a single place among many. There are all `visit***` methods in which we just call `addRoot` with `Many` by default, for example, `visitUnary`, `visitBinary`, ..., basically all the places, where they can be `ExactType`. Again, I'm not insisting we should change all this, I'm just asking... So what you meant by earlier comments is that when we actually use this in GUFA, we don't only query `PossibleContents` but also query the actual type of each expression, so that's fine?

Reply:
We talked about this offline and I think I now understand better about this thanks to @kripken's patient explanation. I'll try to summarize what I understood:

Generally, we don't want to flow redundant info with `PossibleContents` when that info can be supplied from the expression/local/whatever's type. For example, we can query the expression to see if an expression is `i32`, so flowing `ExactType{i32}` is more work and doesn't really add benefits. When we need additional info in `PossibleContents` is when the expression/local/whatever's type is a reference type, in which case it can be exactly that type or one of its subtypes. So we flow `ExactType` only in those cases, and don't flow `ExactType`s for expressions that deal with non-references, such as `visitUnary`, `visitBinary`, or `visitLoad`.

I asked then why we don't rename `ExactType` to something like `ExactRefType`, given that this is only used for references. @kripken's rationale is we want to make `PossibleContents` as general as possible, so even if GUFA doesn't use `ExactType` in `PossibleContents` for references, other analysis that use `PossibleContents` can make use of that feature, so we don't want to remove an existing capability. In the same vein, we want to make `combine` improvements on `ExactType` done in https://github.com/WebAssembly/binaryen/commit/6c5affd482c96e7a1a076fb84fee86391c835b3f, because this makes `PossibleContents` class fully general and its `combine` now does standard type lattice computations.